### PR TITLE
add docs, extract normaliser, helper functions

### PR DESCRIPTION
**Pull request by purarue**
- add iter normalizer, docs, WIP
- spelling fixes

### CODE REVIEW

**Comment by purarue** [comment]
still need to test `iter` a bit more, will migrate some of my sources over to using it

not sure what 'wdir' means, so wasnt able to document that yet

let me know if any of the ideas in the docs are misleading/incorrect, I may be misunderstanding something. I've been using the term 'snapshot' instead of 'cleaned' when describing the logic behind bleanser, and I think snapshot describes what the cleaned file eventually becomes well, but if you have a different term youve been using happy to use that instead

I believe the CI failures will be fixed once #21 is merged, will merge that into here afterwards and then can just squash this when merging

**Comment by purarue** [comment]
Ok yeah, pretty much works as expected:

```diff
diff --git a/src/bleanser_sean/modules/chess.py b/src/bleanser_sean/modules/chess.py
index 1fe695e..acccf01 100644
--- a/src/bleanser_sean/modules/chess.py
+++ b/src/bleanser_sean/modules/chess.py
@@ -1,11 +1,13 @@
-from typing import Iterator, Any, Dict
+import orjson
+from pathlib import Path
+from typing import Iterator, Any
 
-from ..line_normalizer import JsonLineNormalizer
+from bleanser.modules.iter import IterNormaliser
 
 
-class Normalizer(JsonLineNormalizer):
-    @classmethod
-    def handle_json(cls, data: Dict[Any, Any]) -> Iterator[Any]:
+class Normalizer(IterNormaliser):
+    def parse_file(self, file: Path) -> Iterator[Any]:
+        data = orjson.loads(file.read_bytes())
         for b in data:
             # lichess/chess.com keys for time
             assert "createdAt" in b or "end_time" in b, f"missing end time key {b}"
diff --git a/src/bleanser_sean/modules/trakt.py b/src/bleanser_sean/modules/trakt.py
index 4e4b211..2cb9766 100644
--- a/src/bleanser_sean/modules/trakt.py
+++ b/src/bleanser_sean/modules/trakt.py
@@ -1,11 +1,13 @@
-from typing import Iterator, Any, Dict
+import orjson
+from pathlib import Path
+from typing import Iterator, Any
 
-from ..line_normalizer import JsonLineNormalizer
+from bleanser.modules.iter import IterNormaliser
 
 
-class Normalizer(JsonLineNormalizer):
-    @classmethod
-    def handle_json(cls, data: Dict[Any, Any]) -> Iterator[Any]:
+class Normalizer(IterNormaliser):
+    def parse_file(self, path: Path) -> Iterator[Any]:
+        data = orjson.loads(path.read_bytes())
         for b in data["history"]:
             yield b["id"]
```

**Comment by purarue** [comment]
I think might be useful to describe --multiway and --prune-dominated in the help text/docs, as they're sort of bleanser specific and would be useful for the user to know what that means when using it in the CLI. I don't think I can describe those myself so would appreciate you doing so when you can

File: README.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  { "id": 6, "images": [{}], "href": "..." },
+  { "id": 7, "images": [{}], "href": "..." }
+]
+```
+
+When comparing this, you could possibly:
+
+1. Just write the `id` to the file. This is somewhat risky as you don't know if the `href` will always remain the same, so you may be losing data
+1. Write the `id` and the `href`, by specifying those two keys you're interested in
+1. Write the `id` and the `href`, by deleting the `images` key (this is different from 2!)
```
Comment (karlicoss) [inline README.md]: should be 2 and 3?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+from typing import Iterator
+
+from bleanser.core.processor import BaseNormaliser
+
+class Normaliser(BaseNormaliser):
+
+    @contextmanager
+    def do_cleanup(self, path: Path, *, wdir: Path) -> Iterator[Path]:
+        # temporarily decompress if the data is stored as compressed on disk
+        with self.unpacked(path=path, wdir=wdir) as upath:
```
Comment (karlicoss) [inline README.md:56]: hopefully later we can move it to BaseNormaliser so the user doesn't have to think about it

File: src/bleanser/core/processor.py [hunk 3]
```diff
@@ -22,8 +23,44 @@
 from plumbum import local # type: ignore
 
 
+# Mixin for helpers, this lets user subclass nicer
+class NormaliserHelpers:
+    @staticmethod
+    def unique_file_in_tempdir(unpacked_filepath: Path, wdir: Path, *, suffix: Optional[str] = None) -> Path:
```
Comment (karlicoss) [inline src/bleanser/core/processor.py]: nit: best to make keyword only since they are both of the same type and easy to swap by accident

File: src/bleanser/core/processor.py [hunk 4]
```diff
@@ -22,8 +23,44 @@
 from plumbum import local # type: ignore
 
 
+# Mixin for helpers, this lets user subclass nicer
+class NormaliserHelpers:
```
Comment (karlicoss) [inline src/bleanser/core/processor.py]: hmm maybe best to extract regular utility functions? imo mixin usually hints there is an extra state or something like that, and this only has static methods?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+events = set()
+for file in inputs:
+    for line in file:
+        events.add(line)
+# events is now {'A', 'B', 'C', 'D', 'E'}
+```
+
+You might notice that if you removed 'Day 2', you'd still have an accurate backup, and we'd still have all 5 items, but its not obvious you can remove it since none of these are supersets of each other.
+
+`bleanser` is meant to solve this problem in a data agnostic way, so any export can be converted to a unique 'snapshot', and those can be sorted/compared against each other to find redundant data
```
Comment (karlicoss) [inline README.md]: hmm, snapshot feels a bit confusing as well? Usually implies there is some process in time, whereas in this case the export we're processing is already a snapshot. Maybe "normalised representation" or something like that? Would also be consistent with the name for `Normaliser`? 
Could probably also rename `do_cleanup/cleanup` to `do_normalise/normalise`

File: src/bleanser/modules/iter.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    def _emit_history(self, upath: Path, cleaned) -> None:
+        """
+        calls parse_file to extract lines from the unpacked path
+        subclasses should override that to yield some kind of object
+        out to here
+        """
+        with cleaned.open("w") as f:
+            for line in self.parse_file(upath):
+                # newlines may interfere with the diffing, as we sort the lines
```
Comment (karlicoss) [inline src/bleanser/modules/iter.py]: another option could be something like `f.write(repr(str(line)))`? it should make sure there is no weird stuff that ends up on the terminal

File: src/bleanser/modules/iter.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    possible things this could return is a unique key/id, or a tuple of (key, data), or a
+    namedtuple/dataclass
+
+    newlines are stripped from the string, so lines can be compared/diffed properly
+
+    Its possible you could use a library or code from https://github.com/karlicoss/HPI
+    in parse_file, to use the DAL itself to parse the file https://beepb00p.xyz/exports.html#dal
+    """
+
+    def parse_file(self, path: Path) -> Iterator[Any]:
```
Comment (karlicoss) [inline src/bleanser/modules/iter.py]: Maybe a better name would be `ExtractObjectsNormaliser` and `extract_objects` instead of `parse_file` (or something like that)? Feels like it would represent exactly what's happening here?

**Review by karlicoss — CHANGES_REQUESTED** [review]
Thanks, looks great! some mostly minor comments

File: README.md (hunk 1 above)
Comment (purarue) [reply README.md]: markdown automatically renders these as 2 and 3, you dont have to write it out manually. this way if you added something in between 1 and 2 it just updates the numbers automatically

File: src/bleanser/core/processor.py (hunk 4 above)
Comment (purarue) [reply src/bleanser/core/processor.py]: Yeah I was considering both. It does mean more imports everywhere, but might be confusing as a mixin. I think youre right, will just make them standalone functions

**Comment by karlicoss** [comment]
Regarding "dominated" and "multiway" -- I can add proper docs later, but for now explain here!

These tests may be helpful (not sure whether it's easy to follow though)
https://github.com/karlicoss/bleanser/blob/deae59f956ceb1131ed8f8f3666516f63ad82757/src/bleanser/core/processor.py#L798-L860

They are actually used [here](https://github.com/karlicoss/bleanser/blob/master/src/bleanser/core/processor.py#L497-L513) -- but that bit is definitely not easy to follow :see_no_evil: 

Btw there are some (slightly outdated) explanations on Zulip as well https://memex.zulipchat.com/#narrow/stream/279601-hpi/topic/bleanser -- I think this sort of covers both of "dominated" and "multiway"? But let me know if it's still confusing to you.

Overall, these flags are here because they were somewhat experimental (especially multiway mode which is much more complicated than "twoway"), but it likely just makes sense to make MULTIWAY and DELETE_DOMINATED true by default and only set them to false if necessary for some reason. Also twoway/multiway isn't necessarily a good name, so if you can think of something better I'm open for suggestions!

File: README.md (hunk 1 above)
Comment (karlicoss) [reply README.md]: aah I see, didn't know! A bit confusing when you read as markdown (especially when referring to the other items), but not a huge deal!

File: README.md (hunk 5 above)
Comment (purarue) [reply README.md]: yeah, will make a separate issue for that I think, can keep backwards compatability by just having the `do_cleanup` in `BaseNormaliser` call `do_normalize` after sending a warning

**Comment by purarue** [comment]
ok, ended up just adapting the zulip convo because I dont think I could describe it in simpler terms, added that to `doc/`

not sure if there are better names... it feels like such a specific thing that cant find any related terminology

made all the other changes

File: README.md (hunk 2 above)
Comment (purarue) [reply README.md:56]: yeah, this is sort of the most complex case shown in the readme first. when typically the user doesnt have to do this.... 

I think it is still useful for the user to think about it/know what the general steps are. might be confusing to start with, but at least they know whats going on

too much hiding complexity with the base class does make it harder to reason about subclassing

**Comment by karlicoss** [comment]
Awesome, looks great, thanks!
Btw wanna move ObjectsNormaliser to `core/`? Perhaps we could use `bleanser/core/modules/` subpackage (and later move stuff like sqlite/json/xml there?)
The only reason json or xml are in `bleanser/modules` is just happened historically and then I was too lazy to migrate, but the idea with core and modules split is similar to HPI or promnesia otherwise.

**Comment by purarue** [comment]
think I got all the modules, should be good to merge

**Comment by karlicoss** [comment]
3.8 osx pipeline got stuck, but I'm pretty sure it's just github shenanigans, so going to merge. Thanks, might try the object normaliser with whatsapp soon!

**Comment by karlicoss** [comment]
started some work on getting rid of `wdir` and generally cleaner interface

**Comment by karlicoss** [comment]
will do a bit more testing, but I think it's a bit better now! https://github.com/karlicoss/bleanser/pull/26
